csrc/contiguity.cpp

```diff
@@ -87,292 +90,401 @@ bool OrderedIdInformation::checkExclusivelyConsumesAllocs(IterDomain* id) {
-      // However, merging over a broadcast should be fine.
-      inner_pos <= pos_after_outer;
-
-  if (!using_id_graph_) {
-    // TODO: Concretization may prevent contiguous indexing or vectorization.
-    //  It prevents contiguous indexing if the concretization is within the IDs
-    //  that are used for indexing.
-    //  For vectorization it just means we need to make sure the extents of the
-    //  axes to the right of the broadcast allocation domain in the contigous
-    //  merge is bigger than the vectorization dimension. And that the tensor
-    //  buffer supports the vector word size (always done).
-    //
-    // This shouldn't matter when using the IdModel-based
-    // indexer. When concretized, that should be reflected with the
-    // indexing path, and as long as the indexing path has a
-    // contiguous merge, its output should be safe to index. See
-    // also ContigIndexingTest.ConcretizedBroadcastMerge for a
-    // concrete example.
-    bool outer_is_concretized_bcast =
-        merge->outer()->isBroadcast() && isConcretized(merge->outer());
-    bool inner_is_concretized_bcast =
-        merge->inner()->isBroadcast() && isConcretized(merge->inner());
-    out_ordered = out_ordered && !inner_is_concretized_bcast &&
-        !outer_is_concretized_bcast;
-  }
-
-  if (out_ordered) {
-    consistently_ordered_ids_.emplace(merge->out());
-  }
-
-  // Don't just remove active_ids_, as if we have something like:
-  //   [i0, i1, i2, i3]
-  //   ->merge(0, 2)
-  //   ->merge(1)
-  // The latter merge looks like it's ordered correctly, if we update the active
-  // map as:
-  //   [i0, i1, i2, i3] -> [i0*i2, i1, i3]
-  // Hoever if we instead mark it as:
-  //   [i0, i1, i2, i3] -> [i0*i2, i1, nullptr, i3]
-  // Or:
-  //   [i0, i1, i2, i3] -> [nullptr, i1, i0*i2, i3]
-  // It's clear the second merge is not ordered correctly. Doesn't matter which
-  // direction we put the iter domain in, prefer putting it in outer as we often
-  // are looking for inner dimensions that are contiguous. We don't want to
-  // always do this, as it could make ordered merges look non-ordered.
-  // For exmaple: [i0, i1, i2, i3]
-  // ->merge(0)
-  // ->merge(1)
-  // ->merge(0)
-  // If it's updated as:
-  // [i0, i1, i2, i3]
-  // -> [i0*i1, nullptr, i2, i3]
-  // -> [i0*i1, nullptr, i2*i3, nullptr]
-  // Now the final merge looks non-ordered but it is. So only insert a nullptr
-  // entry if the out is not ordered.
-  active_ids_[outer_pos] = merge->out();
-
-  if (!out_ordered) {
-    active_ids_[inner_pos] = nullptr;
+
+    // Update the alloc_id entry for the output.
+    VectorOfUniqueEntries<IterDomain*> alloc_ids = inner_alloc_ids;
+    alloc_ids.pushBack(outer_alloc_ids);
+
+    id_to_alloc_ids_[merge->out()] = alloc_ids;
+
+    // Need to check this after updating active_ids_ and id_to_alloc_ids_
+    if (checkExclusivelyConsumesAllocs(merge->out())) {
+      exclusively_consumes_allocs_.emplace(merge->out());
+    }
```

Comment:
There's no change here. It's just indented due to the if branch at the top of this function.